Optimizer/LexicaseOptimizer/LexicaseOptimizer.cpp

```diff
@@ -0,0 +1,268 @@
+
+	while (keepers.size() > 1 && formulasOrder.size() > 0) {
+		// while there are still atleast one keeper and there are still formulas
+		int formulaIndex = formulasOrder.back();
+		formulasOrder.pop_back();
+
+		double scoreCutoff;
+
+		if (epsilonByRange) {
+			// get the range of scores
+			double maxScoreHere = scores[keepers[0]][formulaIndex];
+			double minScoreHere = maxScoreHere;
+			for (size_t i = 0; i < keepers.size(); i++) {
+				maxScoreHere = std::max(scores[keepers[i]][formulaIndex], maxScoreHere);
+				minScoreHere = std::min(scores[keepers[i]][formulaIndex], minScoreHere);
+			}
+			// get scoreCutoff relivite to min and max
+			scoreCutoff = (maxScoreHere - ((maxScoreHere - minScoreHere) * epsilon));
+		}
+		else {
+			// create a vector of remaning scores and sort them
+			std::vector<double> keeperScores;
+			for (size_t i = 0; i < keepers.size(); i++) {
+				keeperScores.push_back(scores[keepers[i]][formulaIndex]);
+			}
+			std::sort(keeperScores.begin(), keeperScores.end());
+			// scoreCutoff is the (size of vector * epsilon) + 1 from the back
+			scoreCutoff = keeperScores[keeperScores.size()
+				- static_cast<int>(((static_cast<double>(keeperScores.size())) * epsilon) + 1)];
+		}
+		// for each keeper, see if there are still a keeper, i.e. they have score >= scoreCutoff
+		for (size_t i = 0; i < keepers.size();) {
+			if (scores[keepers[i]][formulaIndex] >= scoreCutoff) {
+				i++; // this is a keeper
+			}
+			else {
+				keepers[i] = keepers.back();
+				keepers.pop_back(); // this is not a keeper!
+			}
+		}
+	}
+	int pickHere = Random::getIndex(keepers.size());
+	//std::cout << "    keeping: " << tournamentPopulation[keepers[pickHere]]->ID << std::endl;
+	return keepers[pickHere];
+}
+
+void LexicaseOptimizer::optimize(std::vector<std::shared_ptr<Organism>> &population) {
+	int oldPopulationSize = static_cast<int>(population.size());
+	int nextPopulationTargetSize = nextPopSizeFormula->eval(PT)[0];
+	if (nextPopulationTargetSize == -1) {
+		nextPopulationTargetSize = population.size();
+	}
+
+	int nextPopulationSize = 0;
+
+	// initialize ave,max and, min vectors
+	std::vector<double> aveScores;
+	std::vector<double> maxScores;
+	std::vector<double> minScores;
+	for (size_t fIndex = 0; fIndex < optimizeFormulasMTs.size(); fIndex++) {
+		aveScores.push_back(0);
+		maxScores.push_back(optimizeFormulasMTs[fIndex]->eval(population[0]->dataMap, PT)[0]);
+		minScores.push_back(maxScores.back());
+	}
+
+	killList.clear(); // will hold orgs which will be cleaned up after archive
+	scoresHaveDelta = false; // if all scores are the same, then random selection will result
+
+	// get all scores and record to dataMap
+	scores.clear();
+	for (size_t i = 0; i < population.size(); i++) {
+		// for each org in population
+		killList.insert(population[i]); // add to kill list so that they are deleted in cleanup step
+		scores.push_back({});
+		for (size_t fIndex = 0; fIndex < optimizeFormulasMTs.size(); fIndex++) {
```

Comment:
the min max ave stuff could be factored to use std::min_element std::max_element std::accumulate

Reply:
yes... I agree, but that would change score ... see if_remove above.